diff -ru gtk+-2.11.3/gtk/gtkbuilder.h gtk+-2.11.3/gtk/gtkbuilder.h
--- gtk+-2.11.3/gtk/gtkbuilder.h.typename	2007-06-16 17:36:17.000000000 +0100
+++ gtk+-2.11.3/gtk/gtkbuilder.h	2007-06-16 17:36:42.000000000 +0100
@@ -60,7 +60,7 @@
   GObjectClass parent_class;
   
   GType (* get_type_from_name) (GtkBuilder *builder,
-                                const char *typename);
+                                const char *type_name);
 
   /* Padding for future expansion */
   void (*_gtk_reserved1) (void);
@@ -103,7 +103,7 @@
                                                   const gchar  	*domain);
 const gchar* gtk_builder_get_translation_domain  (GtkBuilder   	*builder);
 GType        gtk_builder_get_type_from_name      (GtkBuilder   	*builder,
-                                                  const char   	*typename);
+                                                  const char   	*type_name);
 
 gboolean     gtk_builder_value_from_string       (GParamSpec   	*pspec,
                                                   const gchar  	*string,
