--- gtk+-2.6.0/gdk-pixbuf/gdk-pixbuf-io.c.lib64	2004-12-21 14:33:41.296252000 -0500
+++ gtk+-2.6.0/gdk-pixbuf/gdk-pixbuf-io.c	2004-12-21 14:35:57.858554000 -0500
@@ -258,7 +258,17 @@
   gchar *result = g_strdup (g_getenv ("GDK_PIXBUF_MODULE_FILE"));
 
   if (!result)
-	  result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gdk-pixbuf.loaders", NULL);
+  {
+	  result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", 
+				     HOST, "gdk-pixbuf.loaders", NULL);
+	  if (!g_file_test (result, G_FILE_TEST_EXISTS))
+	  {
+		  g_free (result);
+		  
+		  result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", 
+					     "gdk-pixbuf.loaders", NULL);
+	  }
+  }
 
   return result;
 }
--- gtk+-2.6.0/gdk-pixbuf/Makefile.am.lib64	2004-12-21 14:33:49.952587000 -0500
+++ gtk+-2.6.0/gdk-pixbuf/Makefile.am	2004-12-21 14:34:45.644839000 -0500
@@ -274,6 +274,7 @@
 	-I$(top_srcdir)/gdk-pixbuf 		\
 	-I$(top_builddir)/gdk-pixbuf 		\
 	-DGTK_SYSCONFDIR=\"$(sysconfdir)\"	\
+	-DHOST=\"$(host_triplet)\"		\
 	-DGTK_VERSION=\"$(GTK_VERSION)\"	\
 	-DGTK_BINARY_VERSION=\"$(GTK_BINARY_VERSION)\"	\
 	-DG_DISABLE_DEPRECATED			\
--- gtk+-2.6.0/gtk/gtkrc.c.lib64	2004-12-21 14:34:02.909617000 -0500
+++ gtk+-2.6.0/gtk/gtkrc.c	2004-12-21 14:36:39.265106000 -0500
@@ -377,7 +377,17 @@
       if (im_module_file)
 	result = g_strdup (im_module_file);
       else
-	result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gtk.immodules", NULL);
+	{
+	  result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", GTK_HOST, 
+				     "gtk.immodules", NULL);
+	  if (!g_file_test (result, G_FILE_TEST_EXISTS))
+	    {
+	      g_free (result);
+	      
+	      result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", 
+					 "gtk.immodules", NULL);
+	    }
+	}
     }
 
 #ifdef G_OS_WIN32
