--- gtk+-2.10.10/gtk/gtknotebook.c.detach	2007-03-12 23:36:03.000000000 -0400
+++ gtk+-2.10.10/gtk/gtknotebook.c	2007-03-12 23:42:11.000000000 -0400
@@ -3926,9 +3926,6 @@
   else  
     g_object_ref_sink (page->menu_label);
 
-  /* child visible will be turned on by switch_page below */
-  gtk_widget_set_child_visible (child, FALSE);
-
   if (notebook->menu)
     gtk_notebook_menu_item_create (notebook,
 				   g_list_find (notebook->children, page));
@@ -3942,6 +3939,10 @@
   if (!notebook->first_tab)
     notebook->first_tab = notebook->children;
 
+  /* child visible will be turned on by switch_page below */
+  if (notebook->cur_page != page)
+    gtk_widget_set_child_visible (child, FALSE);
+
   if (tab_label)
     {
       if (notebook->show_tabs && GTK_WIDGET_VISIBLE (child))
