diff -up gtk+-2.24.8/gtk/gtkcellrendereraccel.c.super gtk+-2.24.8/gtk/gtkcellrendereraccel.c
--- gtk+-2.24.8/gtk/gtkcellrendereraccel.c.super	2011-12-08 08:11:49.933674643 -0500
+++ gtk+-2.24.8/gtk/gtkcellrendereraccel.c	2011-12-08 08:11:58.640674768 -0500
@@ -428,7 +428,6 @@ grab_key_callback (GtkWidget
 				       NULL, NULL, NULL, &consumed_modifiers);
 
   accel_mods = event->state;
-  gdk_keymap_add_virtual_modifiers (gdk_keymap_get_for_display (display), &accel_mods);
 
   accel_key = gdk_keyval_to_lower (event->keyval);
   if (accel_key == GDK_ISO_Left_Tab) 
