diff -up gtk+-2.14.7/gdk/x11/gdkdisplay-x11.c.register-randr-events gtk+-2.14.7/gdk/x11/gdkdisplay-x11.c
--- gtk+-2.14.7/gdk/x11/gdkdisplay-x11.c.register-randr-events	2009-02-13 17:19:55.000000000 -0500
+++ gtk+-2.14.7/gdk/x11/gdkdisplay-x11.c	2009-02-13 17:20:39.000000000 -0500
@@ -195,6 +195,8 @@ gdk_display_open (const gchar *display_n
 
       if ((major == 1 && minor >= 2) || major > 1)
 	  display_x11->have_randr12 = TRUE;
+
+      gdk_x11_register_standard_event_type (display, display_x11->xrandr_event_base, RRNumberEvents);
   }
 #endif
   
