From cdb13d233eb5bc223dd980f120c54f17dea30398 Mon Sep 17 00:00:00 2001
From: Johan Dahlin <johan@gnome.org>
Date: Thu, 31 Dec 2009 00:48:50 +0000
Subject: Make sure Gdk-2.0.gir is installed

Fixes a variable name typo
---
diff --git a/gdk/Makefile.am b/gdk/Makefile.am
index 7dfc193..e1cd8c7 100644
--- a/gdk/Makefile.am
+++ b/gdk/Makefile.am
@@ -211,7 +211,7 @@ girdir = $(datadir)/gir-1.0
 dist_gir_DATA = $(INTROSPECTION_GIRS)
 
 typelibsdir = $(INTROSPECTION_TYPELIBDIR)
-typelibs_DATA = $(BUILT_GIRSOURCES:.gir=.typelib)
+typelibs_DATA = $(INTROSPECTION_GIRS:.gir=.typelib)
 
 CLEANFILES += $(dist_gir_DATA) $(typelibs_DATA)
 endif
--
cgit v0.8.3.1
