diff -up gtk+-2.12.3/gtk-doc.make.rebase gtk+-2.12.3/gtk-doc.make
--- gtk+-2.12.3/gtk-doc.make.rebase	2007-12-18 12:09:09.000000000 -0500
+++ gtk+-2.12.3/gtk-doc.make	2007-12-18 12:11:22.000000000 -0500
@@ -139,9 +139,8 @@ install-data-local:
 	  done; \
 	  echo '-- Installing $(srcdir)/html/index.sgml' ; \
 	  $(INSTALL_DATA) $(srcdir)/html/index.sgml $(DESTDIR)$(TARGET_DIR) || :; \
-	  if test "`which gtkdoc-rebase`" != ""; then \
-        gtkdoc-rebase --relative --dest-dir=$(DESTDIR) --html-dir=$(DESTDIR)$(TARGET_DIR) ; \
-	  fi \
+	  which gtkdoc-rebase > /dev/null && \
+            gtkdoc-rebase --relative --dest-dir=$(DESTDIR) --html-dir=$(DESTDIR)$(TARGET_DIR) ; \
 	fi
 	
 
