diff -up gtk+-2.18.0/gtk/gtkiconview.c.iconview-crash gtk+-2.18.0/gtk/gtkiconview.c
--- gtk+-2.18.0/gtk/gtkiconview.c.iconview-crash	2009-09-28 12:38:46.956098207 -0400
+++ gtk+-2.18.0/gtk/gtkiconview.c	2009-09-28 12:40:03.636858596 -0400
@@ -3469,6 +3469,9 @@ gtk_icon_view_get_item_at_coords (GtkIco
   GList *items, *l;
   GdkRectangle box;
 
+  if (cell_at_pos)
+    *cell_at_pos = NULL;
+
   for (items = icon_view->priv->items; items; items = items->next)
     {
       GtkIconViewItem *item = items->data;
@@ -3503,9 +3506,6 @@ gtk_icon_view_get_item_at_coords (GtkIco
 		    }
 		}
 
-	      if (cell_at_pos)
-		*cell_at_pos = NULL;
-
 	      if (only_in_cell)
 		return NULL;
 	    }
