Index: gdk-pixbuf/io-icns.c
===================================================================
--- gdk-pixbuf/io-icns.c	(revision 20245)
+++ gdk-pixbuf/io-icns.c	(working copy)
@@ -364,7 +364,7 @@
   return pixbuf;
 }
 
-#ifndef INCLUDE_png
+#ifndef INCLUDE_icns
 #define MODULE_ENTRY(function) G_MODULE_EXPORT void function
 #else
 #define MODULE_ENTRY(function) void _gdk_pixbuf__icns_ ## function
Index: gdk-pixbuf/io-ico.c
===================================================================
--- gdk-pixbuf/io-ico.c	(revision 20245)
+++ gdk-pixbuf/io-ico.c	(working copy)
@@ -1184,7 +1184,7 @@
 	return TRUE;
 }
 
-#ifndef INCLUDE_png
+#ifndef INCLUDE_ico
 #define MODULE_ENTRY(function) G_MODULE_EXPORT void function
 #else
 #define MODULE_ENTRY(function) void _gdk_pixbuf__ico_ ## function
Index: gdk-pixbuf/io-jasper.c
===================================================================
--- gdk-pixbuf/io-jasper.c	(revision 20245)
+++ gdk-pixbuf/io-jasper.c	(working copy)
@@ -262,7 +262,7 @@
 	return TRUE;
 }
 
-#ifndef INCLUDE_png
+#ifndef INCLUDE_jasper
 #define MODULE_ENTRY(function) G_MODULE_EXPORT void function
 #else
 #define MODULE_ENTRY(function) void _gdk_pixbuf__jasper_ ## function
