diff -up gtk+-2.11.6/gtk/gtktooltip.c.systray-tooltips gtk+-2.11.6/gtk/gtktooltip.c
--- gtk+-2.11.6/gtk/gtktooltip.c.systray-tooltips	2007-07-25 13:46:43.000000000 -0400
+++ gtk+-2.11.6/gtk/gtktooltip.c	2007-07-25 13:47:05.000000000 -0400
@@ -1116,7 +1116,7 @@ _gtk_tooltip_handle_event (GdkEvent *eve
   /* Hide the tooltip when there's no new tooltip widget */
   if (!has_tooltip_widget)
     {
-      if (current_tooltip && GTK_TOOLTIP_VISIBLE (current_tooltip))
+      if (current_tooltip)
 	gtk_tooltip_hide_tooltip (current_tooltip);
 
       return;
